### ISSUE TITLE
Admin > Countries > Country profile: Add History tab and track changes to disaggregation text

### ISSUE THREAD

**Issue by jennyhavah** [mercycorps/TolaActivity#1873]
# Issue

Country disaggregation text (e.g. "SADD" or "Ethnic group") can be edited at any time, and there are valid reasons to allow this, even after the disaggregation is in use. While our PaQ team advisors do not consider the changes significant enough to require a rationale prompt, they DO consider it significant enough to track in a change log. 

# Acceptance criteria

## Add History tab to Country profiles

![image](https://user-images.githubusercontent.com/33670923/62670009-857e3180-b946-11e9-883c-a36b68fc24a6.png)

- [x] 1. There is a new tab -- the final tab -- in the Country Profile called "History". It is identical in format to the "Status and History" tab in User and Program profiles -- except that we have no concept of Country "status" in Admin Lite, so it's just a "History" tab.)

The only thing in this tab is the change log table.

## Track changes to disaggregation

I propose we track the following "change types" in the change log table. If any of these significantly increase LoE, let's talk. The highest priority item is #3 in the list.

- [x] 1. Country disaggregation added
- [x] 2. Country disaggregation deleted
- [x] 3. Country disaggregation edited  <-- refers to the Disaggregation text field
- [x] 4. Country disaggregation archived <-- see #1850
- [x] 5. Country disaggregation unarchived <-- ditto
- [x] 6. Country disaggregation categories modified
     - [x] 6.1. Previous and Next columns include the disaggregation text (so you know which disaggregation's categories were modified), followed by any categories that were changed in some way -- added, deleted, or edited. All changes saved together should be displayed together in a single row -- just as we do with modifications to targets.

## Housekeeping
* [ ]  Unit test/s written wherever applicable
* [ ]  i18n tags included wherever applicable
* [ ]  Functionality is accessible to people using screen readers

**Comment by sanjuroj** [comment]
@jennyhavah Do you want to track the changes to the sort order of the categories?

**Comment by jennyhavah** [comment]
@sanjuroj If it's easy to do, I say why not. If it would notably increase the level of effort, let me know and I'll run it by MEL.

**Comment by sanjuroj** [comment]
@jennyhavah I'm working on the error handling for the history tab of the Country Admin.  I'm curious how you want to handle errors in fetching history data from the server. This is done when you first expand a country or you make any of the changes that are supposed to get logged.  It's possible that the updates to the country/disaggregation will succeed but the retrieval of history data will fail.  Do you want it to fail silently, since it's possible that they'll never look at the History tab?  Or do you want to provide a history-data-specific message.  If so, what should the text of the message be?  I'm using "Failed to update history data", but that is the thing that made me wonder if we should show a pnotify error at all, because users might be like ????.  Another possibility would be to show an error where the history data would usually appear.

**Comment by jennyhavah** [comment]
@sanjuroj This sounds like a global issue, right? Not specific to Countries? How do we handle this in other History tabs, like the one for Users?

**Comment by sanjuroj** [comment]
@jennyhavah  It is.  The current behavior for the history tabs is that all of the communications with the server are wrapped up in a bundle (i.e. the updating and organization and fetching the new history).  If any of these actions fail, the user sees a generic error message like "Saving failed".  This is true even if, for instance, saving the update succeeds but fetching the history doesn't.  

For the country history tab I've separated out the history query from the other communication with the server, which is why there is the potential to get a separate error message about fetching the history (successful history queries are currently silent).

I could just make the country history the same as the other history tabs.  For some reason, I hadn't considered this, probably because it seemed like a bad idea to allow a failure of the history query look to the user like a failure to save an update.  But it's certainly an option.

**Comment by jennyhavah** [comment]
@sanjuroj I'm a bit confused about the line between system error handling and change tracking, would like to chat about this. 

Is this issue a blocker to completing the Goat Tree specific work on this ticket? Could it go into a separate Poblano ticket that tackles the issue globally?

**Comment by sanjuroj** [comment]
Not really a blocker for Goat Tree, we'll just need to make a decision about how to do it, and the answer may be the same way as the other admin histories do it.  But let's talk about it tomorrow.

**Comment by jennyhavah** [comment]
@sanjuroj That's my inclination as well -- make it consistent with the others for now and file a new ticket to improve across the board.

**Comment by sanjuroj** [comment]
A couple of testing notes:  
1.  As part of this ticket, I refactored the code that displays the change logs across all of the admin types.  It may be a good idea to test the other admin types (programs, users, orgs) to make sure I didn't mess anything up.

2. There is a layout oddity that would be good for someone to confirm.  It takes place when you use the ordering handles to reorder disaggregation labels.  When you've moved the label enough that it overlaps another label, the entire underlying column layout shifts widths (i.e. the "background"  Country, Organizations, Programs, Users columns). i'm going to file this as a separate ticket.

_Not using the reordering handles:_
![Screen Shot 2020-02-03 at 11 24 08 AM](https://user-images.githubusercontent.com/6685499/73683875-eed1c980-4677-11ea-913f-6c88df45d067.png)

_Using the handles:_
![Screen Shot 2020-02-03 at 11 24 18 AM](https://user-images.githubusercontent.com/6685499/73683874-eed1c980-4677-11ea-8644-a3a9f888cb06.png)

**Comment by jennyhavah** [comment]
@sanjuroj History tab looks great, working as expected. 

One text change request: Please change "Disaggregation type" to "Disaggregation" to match the label. (I know we still have "types" on the back end, but we de-jargonized the UI.)

![image](https://user-images.githubusercontent.com/33670923/74198938-9b332300-4c17-11ea-8916-fce5f2ddf32e.png)

**Comment by jennyhavah** [comment]
Just for the record, here are all the different types of disaggregation updates I tried:

![image](https://user-images.githubusercontent.com/33670923/74198996-bdc53c00-4c17-11ea-8ab3-7182bccf622c.png)

**Comment by jennyhavah** [comment]
@sanjuroj The history tab for user changes doesn't seem to be capturing changes as expected. Here's QA:

![image](https://user-images.githubusercontent.com/33670923/74199363-cb2ef600-4c18-11ea-85ae-310ead0cf772.png)

Versus Prod -- created a test user called Test TestTest that we should delete as soon as we're done experimenting:

![image](https://user-images.githubusercontent.com/33670923/74199393-db46d580-4c18-11ea-9609-5e4740aa0eba.png)

**Comment by jennyhavah** [comment]
History tab for Programs looks good.

![image](https://user-images.githubusercontent.com/33670923/74200317-835d9e00-4c1b-11ea-86a5-0389ead2b753.png)

History tab for Organizations looks good.

![image](https://user-images.githubusercontent.com/33670923/74200437-d33c6500-4c1b-11ea-9ac4-8a443eb4d2e9.png)

**Comment by sanjuroj** [comment]
Waiting on someone to look at the layout issue I mentioned in my earlier comment.  "Someone" just means me, Cameron, or Paul, whoever gets to it first.

**Comment by jennyhavah** [comment]
PASSED. User changes issue is fixed.

Layout issue is covered in ticket #1821 

@sanjuroj I don't recall the outcome of the error handling conversation. Can you please create a new ticket if we still need one?

**Comment by sanjuroj** [comment]
@jennyhavah I think we decided to have failed history fetching fail silently, rather than making it seem like the whole saving process has failed, as is done in the other admin pages.

**Comment by jennyhavah** [comment]
@sanjuroj Yes, that sounds right! Is that work done or do we need a ticket for it?
